sycl/plugins/rocm/pi_rocm.cpp

```diff
@@ -0,0 +1,4617 @@
+  // of piextUSMEnqueuePrefetch is detailed in the USM extension
+  assert(flags == 0u);
+
+  try {
+    ScopedContext active(queue->get_context());
+    result = rocm_piEnqueueEventsWait(queue, num_events_in_waitlist,
+                                      events_waitlist, nullptr);
+    if (event) {
+      event_ptr = std::unique_ptr<_pi_event>(
+          _pi_event::make_native(PI_COMMAND_TYPE_MEM_BUFFER_COPY, queue));
+      event_ptr->start();
+    }
+    result = PI_CHECK_ERROR(hipMemPrefetchAsync(
+        ptr, size, queue->get_context()->get_device()->get(),
+        hipStream));
+    if (event) {
+      result = event_ptr->record();
+      *event = event_ptr.release();
+    }
+  } catch (pi_result err) {
+    result = err;
+  }
+  
+  return result;
+}
+
+/// USM: memadvise API to govern behavior of automatic migration mechanisms
+pi_result rocm_piextUSMEnqueueMemAdvise(pi_queue queue, const void *ptr,
+                                        size_t length, pi_mem_advice advice,
+                                        pi_event *event) {
+                                       
+  assert(queue != nullptr);
+  assert(ptr != nullptr);
+  // TODO implement a mapping to hipMemAdvise once the expected behaviour
+  // of piextUSMEnqueueMemAdvise is detailed in the USM extension
+  return rocm_piEnqueueEventsWait(queue, 0, nullptr, event);
+
+ return PI_SUCCESS;
+}
+
+/// API to query information about USM allocated pointers
+/// Valid Queries:
+///   PI_MEM_ALLOC_TYPE returns host/device/shared pi_host_usm value
+///   PI_MEM_ALLOC_BASE_PTR returns the base ptr of an allocation if
+///                         the queried pointer fell inside an allocation.
+///                         Result must fit in void *
+///   PI_MEM_ALLOC_SIZE returns how big the queried pointer's
+///                     allocation is in bytes. Result is a size_t.
+///   PI_MEM_ALLOC_DEVICE returns the pi_device this was allocated against
+///
+/// \param context is the pi_context
+/// \param ptr is the pointer to query
+/// \param param_name is the type of query to perform
+/// \param param_value_size is the size of the result in bytes
+/// \param param_value is the result
+/// \param param_value_ret is how many bytes were written
+pi_result rocm_piextUSMGetMemAllocInfo(pi_context context, const void *ptr,
+                                       pi_mem_info param_name,
+                                       size_t param_value_size,
+                                       void *param_value,
+                                       size_t *param_value_size_ret) {
+  
+  assert(context != nullptr);
+  assert(ptr != nullptr);
+  pi_result result = PI_SUCCESS;
+  hipPointerAttribute_t hipPointerAttributeType;
+
+  try {
+    ScopedContext active(context);
+    switch (param_name) {
+    case PI_MEM_ALLOC_TYPE: {
+      unsigned int value;
+      // do not throw if hipPointerGetAttribute returns hipErrorInvalidValue
+      // TODO hipPointerGetAttribute与CUDA传参不同
```

Comment:
Could you translate this to English, please?

Reply:
That is my negligence and I am sorry about it.